src/mon/NVMeofGwSerialize.h

```diff
@@ -301,50 +320,94 @@ inline  void decode(
   DECODE_FINISH(bl);
 }
 
-inline  void encode(const NvmeGwTimerState& state,  ceph::bufferlist &bl) {
-  ENCODE_START(1, 1, bl);
-  encode((uint32_t)state.data.size(), bl);
-  for (auto &tm_itr:state.data) {
-    encode((uint32_t)tm_itr.first, bl);// encode key
-    uint32_t tick = tm_itr.second.timer_started;
-    uint8_t  val  = tm_itr.second.timer_value;
-    encode(tick, bl);
-    encode(val,  bl);
-    auto endtime  = tm_itr.second.end_time;
-    // Convert the time point to milliseconds since the epoch
-    uint64_t millisecondsSinceEpoch =
+inline  void encode(const NvmeGwTimerState& state,  ceph::bufferlist &bl,
+  uint64_t features) {
+  uint8_t version = 1;
+  if (HAVE_FEATURE(features, NVMEOFHA)) {
+    version = 2;
+  }
+  ENCODE_START(version, version, bl);
+
+  if (version >= 2) {
+    encode((uint32_t)state.data.size(), bl);
+    for (auto &tm_itr:state.data) {
+      encode((uint32_t)tm_itr.first, bl);// encode key
+      uint32_t tick = tm_itr.second.timer_started;
+      uint8_t  val  = tm_itr.second.timer_value;
+      encode(tick, bl);
+      encode(val,  bl);
+      auto endtime  = tm_itr.second.end_time;
+      // Convert the time point to milliseconds since the epoch
+      uint64_t  millisecondsSinceEpoch =
       std::chrono::duration_cast<std::chrono::milliseconds>(
-	endtime.time_since_epoch()).count();
-    encode(millisecondsSinceEpoch , bl);
+      endtime.time_since_epoch()).count();
+      encode(millisecondsSinceEpoch , bl);
+    }
+  }else {
+    encode((uint32_t)MAX_SUPPORTED_ANA_GROUPS, bl);
+    Tmdata tmdata[MAX_SUPPORTED_ANA_GROUPS]; // Constructed objects with defaults
+    for (auto &tm_itr:state.data) {
+      tmdata[tm_itr.first].timer_started = tm_itr.second.timer_started;
+      tmdata[tm_itr.first].timer_value = tm_itr.second.timer_value;
+      tmdata[tm_itr.first].end_time = tm_itr.second.end_time;
```

Comment:
This is dangerous.  It sort of works at the moment due to the guard in cfg_add_gw, but during the evolution of this PR, MAX_SUPPORTED_ANA_GROUPS has generally been considered to be for backward compatibility rather than a permanent commitment.  This encoder should really be robust to a future where more than 16 GWs are supported, and it would be a very easy thing to miss while making such a change.  Rather than populating a fixed size array, simply check state.data for each loop through the following for loop.  The branch I mention above has such a fix.

Reply:
If you ever do feel you need such a fixed size array, please be certain to include ceph_assert guards to ensure that the bounds are appropriate if they cannot be statically proven to be so within a screen length or so of the usage.  While crashing the mon is unacceptable, it's actually far preferable to the kind of undefined nonsense that can happen as a result of an out-of-bounds array write.